### DISCUSSION TITLE
Configuring Mp3_Ack_API_Test.ino

### DISCUSSION THREAD

**Discussion by springknees** [Makuna/DFMiniMp3#135] [General]
I'm successfully running a DFPlayer with an Arduino Uno (with Makuna's library) connected via software serial on Pins 10 & 11. Out of academic interest I want to run Debug (Mp3_Ack_API_Test.ino) on my two different types of DFPlayer.

I'm not really up-to-speed with OOP, so I'm in a muddle in my attempts to get Mp3_Ack_API_Test.ino to compile successfully! Can someone please confirm that I should be able to get this to run on my Uno?

If it should compile, I suspect that my problems are due to my inability to correctly comment in/out the various lines that relate to hardware/software serial - using my little knowledge I've tried various permutations without success. If anyone can point me in the right direction it will be much appreciated.

**Comment by Makuna** [comment]
It should work, and the UNO only has one hardware serial (for debug) so you have to switch the sketch to use software serial.

key changes in order that they show up in the file are ...
```
#include <SoftwareSerial.h>

SoftwareSerial Mp3Serial(10, 11); // RX, TX // comment out the "define Mp3Serial Serial1"

typedef DFMiniMp3<SoftwareSerial, Mp3Notify> DfMp3; // comment out the other

mp3.begin(); // no arguments
```

Let me know how that goes.

**Comment by Makuna** [comment]
I updated this sample to cleanup the locations to make it a little more obvious what to do.  So, you can just get/copy that file and make the minor changes.

**Reply by springknees** [reply]
Many thanks Michael - what phenomenally fast support!! I'll get onto this in a little while and report back.

**Comment by springknees** [comment]
Hi - I've now copied the revised example and edited like this:-
```
#define MP3_RX_PIN 10  //Replaced with my pin numbers
#define MP3_TX_PIN 11
```
Uncommented:-
```
#include <SoftwareSerial.h>
SoftwareSerial Mp3Serial(MP3_RX_PIN, MP3_TX_PIN); // RX, TX
#define Mp3SerialType SoftwareSerial
```
Commented:-
```
//#define Mp3SerialType HardwareSerial
//#define Mp3Serial Serial1
```
Changed Baud rate:-

`` DebugOut.begin(9600);``

Uncommented:-

`mp3.begin(); // when you can't set the pins OR software serial is used`

Commented:-

`// mp3.begin(MP3_RX_PIN, MP3_TX_PIN); // rx, tx`

=== The Sketch now compiles and runs = THANKS A LOT FOR YOUR HELP ===

As I said, the player works OK in my limited application, however you might be interested in the result of my test! The SD Card has two Folders ADVERT and mp3, both have six files approx size 5K named 0001.mp3 -> 0006.mp3

The DFPlayer is marked MP3-TF-16P with chip MH2024K-24SS - here's the debug output:-


OUT  7e ff 06 0c 01 00 00 fe ee ef
IN  7e ff 06 41 00 00 00 fe ba ef
IN  7e ff 06 3f 00 00 02 fe ba ef
initializing...
SD Card, online
starting...
OUT  7e ff 06 46 00 00 00 fe b5 ef
IN  7e ff 06 46 00 00 08 fe ad ef
version 8
playGlobalTrack 6
OUT  7e ff 06 03 01 00 06 fe f1 ef
IN  7e ff 06 41 00 00 00 fe ba ef
playMp3FolderTrack 2
OUT  7e ff 06 12 01 00 02 fe e6 ef
IN  7e ff 06 41 00 00 00 fe ba ef
playFolderTrack 20, 2
OUT  7e ff 06 0f 01 14 02 fe d5 ef
IN  7e ff 06 41 00 00 00 fe ba ef
playFolderTrack16 10, 2000
IN  7e ff 06 40 00 00 06 fe b5 ef
OUT  7e ff 06 14 01 a7 d0 fd 6f ef
IN  7e ff 06 41 00 00 00 fe ba ef
playRandomTrackFromAll

Com Error 6
IN  7e ff 06 40 00 00 06 fe b5 ef
OUT  7e ff 06 18 01 00 00 fe e2 ef
IN  7e ff 06 41 00 00 00 fe ba ef
nextTrack

Com Error 6
OUT  7e ff 06 01 01 00 00 fe f9 ef
IN  7e ff 06 41 00 00 00 fe ba ef
prevTrack
OUT  7e ff 06 02 01 00 00 fe f8 ef
IN  7e ff 06 41 00 00 00 fe ba ef
OUT  7e ff 06 4c 00 00 00 fe af ef
IN  7e ff 06 4c 00 00 04 fe ab ef
current track 4
setVolume 16
OUT  7e ff 06 06 01 00 10 fe e4 ef
IN  7e ff 06 41 00 00 00 fe ba ef
OUT  7e ff 06 43 00 00 00 fe b8 ef
IN  7e ff 06 43 00 00 10 fe a8 ef
volume 16
increaseVolume
OUT  7e ff 06 04 01 00 00 fe f6 ef
IN  7e ff 06 41 00 00 00 fe ba ef
decreaseVolume
OUT  7e ff 06 05 01 00 00 fe f5 ef
IN  7e ff 06 41 00 00 00 fe ba ef
loopGlobalTrack 6
IN  7e ff 06 3d 00 00 04 fe ba ef
IN  7e ff 06 3d 00 00 04 fe ba ef
OUT  7e ff 06 08 01 00 06 fe ec ef
IN  7e ff 06 41 00 00 00 fe ba ef
loopFolder 20
Play finished for #4
Play finished for #4
OUT  7e ff 06 17 01 00 14 fe cf ef
IN  7e ff 06 41 00 00 00 fe ba ef
setPlaybackMode 2 (single repeat)
IN  7e ff 06 40 00 00 06 fe b5 ef
OUT  7e ff 06 08 01 00 02 fe f0 ef
IN  7e ff 06 41 00 00 00 fe ba ef

Com Error 6
OUT  7e ff 06 45 00 00 00 fe b6 ef
IN  7e ff 06 45 00 00 02 fe b4 ef
playback mode 2
setRepeatPlayAllInRoot false
OUT  7e ff 06 11 01 00 00 fe e9 ef
IN  7e ff 06 41 00 00 00 fe ba ef
setRepeatPlayCurrentTrack true
OUT  7e ff 06 19 01 00 00 fe e1 ef
IN  7e ff 06 41 00 00 00 fe ba ef
setEq 3 (jazz)
OUT  7e ff 06 07 01 00 03 fe f0 ef
IN  7e ff 06 41 00 00 00 fe ba ef
OUT  7e ff 06 44 00 00 00 fe b7 ef
IN  7e ff 06 44 00 00 03 fe b4 ef
eq mode 3
setPlaybackSource 2 (SD)
OUT  7e ff 06 09 01 00 02 fe ef ef
IN  7e ff 06 41 00 00 00 fe ba ef
start
OUT  7e ff 06 0d 01 00 00 fe ed ef
IN  7e ff 06 41 00 00 00 fe ba ef
pause
OUT  7e ff 06 0e 01 00 00 fe ec ef
IN  7e ff 06 41 00 00 00 fe ba ef
stop
OUT  7e ff 06 16 01 00 00 fe e4 ef
IN  7e ff 06 41 00 00 00 fe ba ef
OUT  7e ff 06 42 00 00 00 fe b9 ef
IN  7e ff 06 42 00 02 00 fe b7 ef
status source is 2 and state is 0
OUT  7e ff 06 4e 00 00 00 fe ad ef
IN  7e ff 06 4e 00 00 00 fe ad ef
folder 10 files 0
OUT  7e ff 06 48 00 00 00 fe b3 ef
IN  7e ff 06 48 00 00 0c fe a7 ef
total files 12
OUT  7e ff 06 4f 00 00 00 fe ac ef
IN  7e ff 06 4f 00 00 03 fe a9 ef
total folders 3
playAdvertisement 2
OUT  7e ff 06 13 01 00 02 fe e5 ef
IN  7e ff 06 41 00 00 00 fe ba ef
stopAdvertisement
IN  7e ff 06 40 00 00 07 fe b4 ef
OUT  7e ff 06 15 01 00 00 fe e5 ef
IN  7e ff 06 41 00 00 00 fe ba ef
enableDac

Com Error 7
OUT  7e ff 06 1a 01 00 00 fe e0 ef
IN  7e ff 06 41 00 00 00 fe ba ef
disableDac
OUT  7e ff 06 1a 01 00 01 fe df ef
IN  7e ff 06 41 00 00 00 fe ba ef
...stopping

**Reply by Makuna** [reply]
Every thing looks great.  

- The Com Error 6 is due to the file/folder requested isn't present and are expected.  Due to the response being asynchronous and really not much of delay between the calls, they may seem a little out of order but this is normal.
- The Com Error 7 is due to not playing an advertisement at the time the call is made, also normal/expected.
- The only strange one here is that it reports that there is THREE folders to play from, but you stated you only had two ADVERTS and MP3.  Could you confirm your folder heirarchy on the SD?  Is there a folder in a folder?

**Comment by springknees** [comment]
Thanks for the feedback - glad to hear that all looks great! I wondered what the Com errors related to. As far as I can see on my Windows system (with hidden files set to show) I only have the two folders ADVERTS and mp3.

I have run the same debug with the same SD card (and same hardware/software) on my other DFPlayer module - it's labelled DFPlayer Mini and has the GD3200B chip - here is the output. You'll note that there are no Com errors!

Please don't waste any time on this, as the player(s) are working fine in my application, however, I'm guessing that analysis may be of interest to you :)

OUT  7e ff 06 0c 01 00 00 fe ee ef
IN  7e ff 06 41 00 00 00 fe ba ef
IN  7e ff 06 3f 00 00 02 fe ba ef
initializing...
SD Card, online
starting...
OUT  7e ff 06 46 00 00 00 fe b5 ef
OUT  7e ff 06 46 00 00 00 fe b5 ef
OUT  7e ff 06 46 00 00 00 fe b5 ef
version 0
playGlobalTrack 6
OUT  7e ff 06 03 01 00 06 fe f1 ef
IN  7e ff 06 41 00 00 00 fe ba ef
playMp3FolderTrack 2
OUT  7e ff 06 12 01 00 02 fe e6 ef
IN  7e ff 06 41 00 00 00 fe ba ef
playFolderTrack 20, 2
OUT  7e ff 06 0f 01 14 02 fe d5 ef
IN  7e ff 06 41 00 00 00 fe ba ef
playFolderTrack16 10, 2000
OUT  7e ff 06 14 01 a7 d0 fd 6f ef
IN  7e ff 06 41 00 00 00 fe ba ef
playRandomTrackFromAll
OUT  7e ff 06 18 01 00 00 fe e2 ef
IN  7e ff 06 41 00 00 00 fe ba ef
nextTrack
OUT  7e ff 06 01 01 00 00 fe f9 ef
IN  7e ff 06 41 00 00 00 fe ba ef
prevTrack
OUT  7e ff 06 02 01 00 00 fe f8 ef
IN  7e ff 06 41 00 00 00 fe ba ef
OUT  7e ff 06 4c 00 00 00 fe af ef
IN  7e ff 06 4c 00 00 08 fe a7 ef
current track 8
setVolume 16
OUT  7e ff 06 06 01 00 10 fe e4 ef
IN  7e ff 06 41 00 00 00 fe ba ef
OUT  7e ff 06 43 00 00 00 fe b8 ef
IN  7e ff 06 43 00 00 10 fe a8 ef
volume 16
increaseVolume
OUT  7e ff 06 04 01 00 00 fe f6 ef
IN  7e ff 06 41 00 00 00 fe ba ef
decreaseVolume
OUT  7e ff 06 05 01 00 00 fe f5 ef
IN  7e ff 06 41 00 00 00 fe ba ef
loopGlobalTrack 6
OUT  7e ff 06 08 01 00 06 fe ec ef
IN  7e ff 06 41 00 00 00 fe ba ef
loopFolder 20
OUT  7e ff 06 17 01 00 14 fe cf ef
IN  7e ff 06 41 00 00 00 fe ba ef
setPlaybackMode 2 (single repeat)
OUT  7e ff 06 08 01 00 02 fe f0 ef
IN  7e ff 06 41 00 00 00 fe ba ef
OUT  7e ff 06 45 00 00 00 fe b6 ef
IN  7e ff 06 40 00 00 03 fe b8 ef
OUT  7e ff 06 45 00 00 00 fe b6 ef
IN  7e ff 06 45 00 00 02 fe b4 ef
playback mode 2
setRepeatPlayAllInRoot false
OUT  7e ff 06 11 01 00 00 fe e9 ef
IN  7e ff 06 41 00 00 00 fe ba ef
setRepeatPlayCurrentTrack true
OUT  7e ff 06 19 01 00 00 fe e1 ef
IN  7e ff 06 41 00 00 00 fe ba ef
setEq 3 (jazz)
OUT  7e ff 06 07 01 00 03 fe f0 ef
IN  7e ff 06 41 00 00 00 fe ba ef
OUT  7e ff 06 44 00 00 00 fe b7 ef
IN  7e ff 06 44 00 00 03 fe b4 ef
eq mode 3
setPlaybackSource 2 (SD)
OUT  7e ff 06 09 01 00 02 fe ef ef
IN  7e ff 06 41 00 00 00 fe ba ef
start
OUT  7e ff 06 0d 01 00 00 fe ed ef
IN  7e ff 06 41 00 00 00 fe ba ef
pause
OUT  7e ff 06 0e 01 00 00 fe ec ef
IN  7e ff 06 41 00 00 00 fe ba ef
stop
OUT  7e ff 06 16 01 00 00 fe e4 ef
IN  7e ff 06 41 00 00 00 fe ba ef
OUT  7e ff 06 42 00 00 00 fe b9 ef
IN  7e ff 06 42 00 00 00 fe b9 ef
status source is 0 and state is 0
OUT  7e ff 06 4e 00 00 00 fe ad ef
IN  7e ff 06 4e 00 00 06 fe a7 ef
folder 10 files 6
OUT  7e ff 06 48 00 00 00 fe b3 ef
IN  7e ff 06 48 00 00 0c fe a7 ef
total files 12
OUT  7e ff 06 4f 00 00 00 fe ac ef
IN  7e ff 06 4f 00 00 02 fe aa ef
total folders 2
playAdvertisement 2
OUT  7e ff 06 13 01 00 02 fe e5 ef
IN  7e ff 06 41 00 00 00 fe ba ef
stopAdvertisement
OUT  7e ff 06 15 01 00 00 fe e5 ef
IN  7e ff 06 41 00 00 00 fe ba ef
enableDac
IN  7e ff 06 3d 00 00 05 fe b9 ef
OUT  7e ff 06 1a 01 00 00 fe e0 ef
IN  7e ff 06 41 00 00 00 fe ba ef
disableDac
Play finished for #5
OUT  7e ff 06 1a 01 00 01 fe df ef
IN  7e ff 06 41 00 00 00 fe ba ef
...stopping

**Reply by Makuna** [reply]
They weren't really comm errors. it's the error returned from the device.  
```
    // from device                
    DfMp3_Error_Busy = 1,         //  busy                  busy
    DfMp3_Error_Sleeping,         //  frame not received    sleep
    DfMp3_Error_SerialWrongStack, //  verification error    frame not received
    DfMp3_Error_CheckSumNotMatch, //                        checksum
    DfMp3_Error_FileIndexOut,     //  folder out of scope   track out of scope
    DfMp3_Error_FileMismatch,     //  folder not found      track not found
    DfMp3_Error_Advertise,        //                        only allowed while playing     advertisement not allowed
    DfMp3_Error_SdReadFail,       //                        SD card failed
    DfMp3_Error_FlashReadFail,    //                        Flash mem failed
    DfMp3_Error_EnteredSleep = 10, //                        entered sleep   
```
It's VERY interesting.  I had a conversation thread that I asked specifically for people to try this and do exactly what you are doing, post the results.

This one doesn't support the get software version call, it just ignores it and thus the version 0.

The missing com errors could be that you had folders and files present that match the request.  That or it just ignores invalid folder files requests.

When calling setPlaybackMode 2, it had a communications error and returned an error for it, the library retried, and it then got a valid answer.  So, it does support knowing the request got corrupted and let the library know so it can retry.

getStatus() returns an empty value set, expected it to return 2/0 (SD and IDLE).  But it returned 0/0 (general and idle).  There are known inconsistencies on this call between chips.


BUT, I noticed this now in both of your captures above...
```
OUT 7e ff 06 4e 00 00 00 fe ad ef
IN 7e ff 06 4e 00 00 06 fe a7 ef
folder 10 files 6
```
The request for folder 10 file count using 
```
uint16_t count = mp3.getFolderTrackCount(10);
```
The debug data doesn't match the call.  
this `OUT 7e ff 06 4e 00 00 00 fe ad ef` should be `OUT 7e ff 06 4e 00 00 0a fe ad ef` in both cases!?
What microcontroller are you using?

**Reply by Makuna** [reply]
well I guess I forgot, but I fixed getFolerTrackTrack() but they fix was in a released version.  A new release was just generated v 1.2.1

**Comment by springknees** [comment]
Hi - just ran two more debugs using library version 1.2.1 - the outputs follow.

PLATFORM: Atmel AVR (4.1.0) > Arduino Uno
HARDWARE: ATMEGA328P 16MHz, 2KB RAM, 31.50KB Flash

SD Card remains as before with 2 Folders, each folder contains six files. If you'd like me to reconfigure the contents of the SD card, let me know. I can also test using an alternative SD card, if required.

**MH2024K-24SS - Module**

OUT  7e ff 06 0c 01 00 00 fe ee ef
IN  7e ff 06 41 00 00 00 fe ba ef
IN  7e ff 06 3f 00 00 02 fe ba ef
initializing...
SD Card, online
starting...
OUT  7e ff 06 46 00 00 00 fe b5 ef
IN  7e ff 06 46 00 00 08 fe ad ef
version 8
playGlobalTrack 6
OUT  7e ff 06 03 01 00 06 fe f1 ef
IN  7e ff 06 41 00 00 00 fe ba ef
playMp3FolderTrack 2
OUT  7e ff 06 12 01 00 02 fe e6 ef
IN  7e ff 06 41 00 00 00 fe ba ef
playFolderTrack 20, 2
OUT  7e ff 06 0f 01 14 02 fe d5 ef
IN  7e ff 06 41 00 00 00 fe ba ef
playFolderTrack16 10, 2000
IN  7e ff 06 40 00 00 06 fe b5 ef
OUT  7e ff 06 14 01 a7 d0 fd 6f ef
IN  7e ff 06 41 00 00 00 fe ba ef
playRandomTrackFromAll

Com Error 6
IN  7e ff 06 40 00 00 06 fe b5 ef
OUT  7e ff 06 18 01 00 00 fe e2 ef
IN  7e ff 06 41 00 00 00 fe ba ef
nextTrack

Com Error 6
OUT  7e ff 06 01 01 00 00 fe f9 ef
IN  7e ff 06 41 00 00 00 fe ba ef
prevTrack
OUT  7e ff 06 02 01 00 00 fe f8 ef
IN  7e ff 06 41 00 00 00 fe ba ef
OUT  7e ff 06 4c 00 00 00 fe af ef
IN  7e ff 06 4c 00 00 0a fe a5 ef
current track 10
setVolume 16
OUT  7e ff 06 06 01 00 10 fe e4 ef
IN  7e ff 06 41 00 00 00 fe ba ef
OUT  7e ff 06 43 00 00 00 fe b8 ef
IN  7e ff 06 43 00 00 10 fe a8 ef
volume 16
increaseVolume
OUT  7e ff 06 04 01 00 00 fe f6 ef
IN  7e ff 06 41 00 00 00 fe ba ef
decreaseVolume
OUT  7e ff 06 05 01 00 00 fe f5 ef
IN  7e ff 06 41 00 00 00 fe ba ef
loopGlobalTrack 6
OUT  7e ff 06 08 01 00 06 fe ec ef
IN  7e ff 06 3d 00 00 0a fe b4 ef
IN  7e ff 06 3d 00 00 0a fe b4 ef
IN  7e ff 06 41 00 00 00 fe ba ef
loopFolder 20
Play finished for #10
Play finished for #10
OUT  7e ff 06 17 01 00 14 fe cf ef
IN  7e ff 06 41 00 00 00 fe ba ef
setPlaybackMode 2 (single repeat)
IN  7e ff 06 40 00 00 06 fe b5 ef
OUT  7e ff 06 08 01 00 02 fe f0 ef
IN  7e ff 06 41 00 00 00 fe ba ef

Com Error 6
OUT  7e ff 06 45 00 00 00 fe b6 ef
IN  7e ff 06 45 00 00 02 fe b4 ef
playback mode 2
setRepeatPlayAllInRoot false
OUT  7e ff 06 11 01 00 00 fe e9 ef
IN  7e ff 06 41 00 00 00 fe ba ef
setRepeatPlayCurrentTrack true
OUT  7e ff 06 19 01 00 00 fe e1 ef
IN  7e ff 06 41 00 00 00 fe ba ef
setEq 3 (jazz)
OUT  7e ff 06 07 01 00 03 fe f0 ef
IN  7e ff 06 41 00 00 00 fe ba ef
OUT  7e ff 06 44 00 00 00 fe b7 ef
IN  7e ff 06 44 00 00 03 fe b4 ef
eq mode 3
setPlaybackSource 2 (SD)
OUT  7e ff 06 09 01 00 02 fe ef ef
IN  7e ff 06 41 00 00 00 fe ba ef
start
OUT  7e ff 06 0d 01 00 00 fe ed ef
IN  7e ff 06 41 00 00 00 fe ba ef
pause
OUT  7e ff 06 0e 01 00 00 fe ec ef
IN  7e ff 06 41 00 00 00 fe ba ef
stop
OUT  7e ff 06 16 01 00 00 fe e4 ef
IN  7e ff 06 41 00 00 00 fe ba ef
OUT  7e ff 06 42 00 00 00 fe b9 ef
IN  7e ff 06 42 00 02 00 fe b7 ef
status source is 2 and state is 0
OUT  7e ff 06 4e 00 00 0a fe a3 ef
IN  7e ff 06 40 00 00 06 fe b5 ef
OUT  7e ff 06 4e 00 00 0a fe a3 ef
IN  7e ff 06 40 00 00 06 fe b5 ef
OUT  7e ff 06 4e 00 00 0a fe a3 ef
IN  7e ff 06 40 00 00 06 fe b5 ef

Com Error 6
folder 10 files 0
OUT  7e ff 06 48 00 00 00 fe b3 ef
IN  7e ff 06 48 00 00 0c fe a7 ef
total files 12
OUT  7e ff 06 4f 00 00 00 fe ac ef
IN  7e ff 06 4f 00 00 03 fe a9 ef
total folders 3
playAdvertisement 2
OUT  7e ff 06 13 01 00 02 fe e5 ef
IN  7e ff 06 41 00 00 00 fe ba ef
stopAdvertisement
IN  7e ff 06 40 00 00 07 fe b4 ef
OUT  7e ff 06 15 01 00 00 fe e5 ef
IN  7e ff 06 41 00 00 00 fe ba ef
enableDac

Com Error 7
OUT  7e ff 06 1a 01 00 00 fe e0 ef
IN  7e ff 06 41 00 00 00 fe ba ef
disableDac
OUT  7e ff 06 1a 01 00 01 fe df ef
IN  7e ff 06 41 00 00 00 fe ba ef
...stopping

**GD3200B - Module**

OUT  7e ff 06 0c 01 00 00 fe ee ef
IN  7e ff 06 41 00 00 00 fe ba ef
IN  7e ff 06 3f 00 00 02 fe ba ef
initializing...
SD Card, online
starting...
OUT  7e ff 06 46 00 00 00 fe b5 ef
OUT  7e ff 06 46 00 00 00 fe b5 ef
OUT  7e ff 06 46 00 00 00 fe b5 ef
version 0
playGlobalTrack 6
OUT  7e ff 06 03 01 00 06 fe f1 ef
IN  7e ff 06 41 00 00 00 fe ba ef
playMp3FolderTrack 2
OUT  7e ff 06 12 01 00 02 fe e6 ef
IN  7e ff 06 41 00 00 00 fe ba ef
playFolderTrack 20, 2
OUT  7e ff 06 0f 01 14 02 fe d5 ef
IN  7e ff 06 41 00 00 00 fe ba ef
playFolderTrack16 10, 2000
OUT  7e ff 06 14 01 a7 d0 fd 6f ef
IN  7e ff 06 41 00 00 00 fe ba ef
playRandomTrackFromAll
OUT  7e ff 06 18 01 00 00 fe e2 ef
IN  7e ff 06 41 00 00 00 fe ba ef
nextTrack
OUT  7e ff 06 01 01 00 00 fe f9 ef
IN  7e ff 06 41 00 00 00 fe ba ef
prevTrack
OUT  7e ff 06 02 01 00 00 fe f8 ef
IN  7e ff 06 41 00 00 00 fe ba ef
OUT  7e ff 06 4c 00 00 00 fe af ef
IN  7e ff 06 4c 00 00 0a fe a5 ef
current track 10
setVolume 16
OUT  7e ff 06 06 01 00 10 fe e4 ef
IN  7e ff 06 41 00 00 00 fe ba ef
OUT  7e ff 06 43 00 00 00 fe b8 ef
IN  7e ff 06 43 00 00 10 fe a8 ef
volume 16
increaseVolume
OUT  7e ff 06 04 01 00 00 fe f6 ef
IN  7e ff 06 41 00 00 00 fe ba ef
decreaseVolume
OUT  7e ff 06 05 01 00 00 fe f5 ef
OUT  7e ff 06 05 01 00 00 fe f5 ef
IN  7e ff 06 40 00 00 03 fe b8 ef
OUT  7e ff 06 05 01 00 00 fe f5 ef
IN  7e ff 06 41 00 00 00 fe ba ef
loopGlobalTrack 6
OUT  7e ff 06 08 01 00 06 fe ec ef
IN  7e ff 06 41 00 00 00 fe ba ef
loopFolder 20
OUT  7e ff 06 17 01 00 14 fe cf ef
IN  7e ff 06 41 00 00 00 fe ba ef
setPlaybackMode 2 (single repeat)
OUT  7e ff 06 08 01 00 02 fe f0 ef
IN  7e ff 06 41 00 00 00 fe ba ef
OUT  7e ff 06 45 00 00 00 fe b6 ef
IN  7e ff 06 40 00 00 03 fe b8 ef
OUT  7e ff 06 45 00 00 00 fe b6 ef
IN  7e ff 06 45 00 00 02 fe b4 ef
playback mode 2
setRepeatPlayAllInRoot false
OUT  7e ff 06 11 01 00 00 fe e9 ef
IN  7e ff 06 41 00 00 00 fe ba ef
setRepeatPlayCurrentTrack true
OUT  7e ff 06 19 01 00 00 fe e1 ef
IN  7e ff 06 41 00 00 00 fe ba ef
setEq 3 (jazz)
OUT  7e ff 06 07 01 00 03 fe f0 ef
IN  7e ff 06 41 00 00 00 fe ba ef
OUT  7e ff 06 44 00 00 00 fe b7 ef
IN  7e ff 06 44 00 00 03 fe b4 ef
eq mode 3
setPlaybackSource 2 (SD)
OUT  7e ff 06 09 01 00 02 fe ef ef
IN  7e ff 06 41 00 00 00 fe ba ef
start
OUT  7e ff 06 0d 01 00 00 fe ed ef
IN  7e ff 06 41 00 00 00 fe ba ef
pause
OUT  7e ff 06 0e 01 00 00 fe ec ef
IN  7e ff 06 41 00 00 00 fe ba ef
stop
OUT  7e ff 06 16 01 00 00 fe e4 ef
IN  7e ff 06 41 00 00 00 fe ba ef
OUT  7e ff 06 42 00 00 00 fe b9 ef
IN  7e ff 06 42 00 00 00 fe b9 ef
status source is 0 and state is 0
OUT  7e ff 06 4e 00 00 0a fe a3 ef
IN  7e ff 06 4e 00 00 06 fe a7 ef
folder 10 files 6
OUT  7e ff 06 48 00 00 00 fe b3 ef
IN  7e ff 06 48 00 00 0c fe a7 ef
total files 12
OUT  7e ff 06 4f 00 00 00 fe ac ef
IN  7e ff 06 4f 00 00 02 fe aa ef
total folders 2
playAdvertisement 2
OUT  7e ff 06 13 01 00 02 fe e5 ef
IN  7e ff 06 41 00 00 00 fe ba ef
stopAdvertisement
OUT  7e ff 06 15 01 00 00 fe e5 ef
IN  7e ff 06 41 00 00 00 fe ba ef
enableDac
IN  7e ff 06 3d 00 00 05 fe b9 ef
OUT  7e ff 06 1a 01 00 00 fe e0 ef
IN  7e ff 06 41 00 00 00 fe ba ef
disableDac
Play finished for #5
OUT  7e ff 06 1a 01 00 01 fe df ef
IN  7e ff 06 41 00 00 00 fe ba ef
...stopping

**Reply by springknees** [reply]
I added two more folders to the SD Card = 4. The MH2024K-24SS now reports Total Folders = 5.

My Win 10 File Explorer is set to Show Hidden Files/Folders - only 4 Folders are listed.

I have now discovered a handy Chrome facility where, by entering  "file:///x:" (the x should be the drive letter of the SD card) in the Address Bar + ENTER, the card contents are displayed. Guess what - there's the extra folder seen by the MH2024K-24SS but not the GD3200B!
![sd_files](https://github.com/Makuna/DFMiniMp3/assets/13821909/fb5e645a-0670-4983-b8bd-2f7b43782620)

**Reply by Makuna** [reply]
System File, not just hidden, but also make sure you see system files.  I know Windows CMD prompt, the DIR command has the option in its attributes.
```
/A          Displays files with specified attributes.
  attributes   D  Directories                R  Read-only files
               H  Hidden files               A  Files ready for archiving
               S  System files               I  Not content indexed files
               L  Reparse Points             O  Offline files
```
